maven-plugin-report-plugin/src/main/java/org/apache/maven/plugin/plugin/report/GoalRenderer.java

```diff
@@ -0,0 +1,533 @@
+        // split into type arguments and main type
+        int startTypeArguments = type.indexOf('<');
+        if (startTypeArguments == -1) {
+            return getShortTypeOfSimpleType(type);
+        } else {
+            StringBuilder shortType = new StringBuilder();
+            shortType.append(getShortTypeOfSimpleType(type.substring(0, startTypeArguments)));
+            shortType
+                    .append("<")
+                    .append(getShortTypeOfTypeArgument(type.substring(startTypeArguments + 1, type.lastIndexOf(">"))))
+                    .append(">");
+            return shortType.toString();
+        }
+    }
+
+    private static String getShortTypeOfTypeArgument(String type) {
+        String[] typeArguments = type.split(",\\s*");
+        StringBuilder shortType = new StringBuilder();
+        for (int i = 0; i < typeArguments.length; i++) {
+            String typeArgument = typeArguments[i];
+            if (typeArgument.contains("<")) {
+                // nested type arguments lead to ellipsis
+                return "...";
+            } else {
+                shortType.append(getShortTypeOfSimpleType(typeArgument));
+                if (i < typeArguments.length - 1) {
+                    shortType.append(",");
+                }
+            }
+        }
+        return shortType.toString();
+    }
+
+    private static String getShortTypeOfSimpleType(String type) {
+        int index = type.lastIndexOf('.');
+        return type.substring(index + 1);
+    }
+
+    private Map.Entry<String, Optional<String>> getLinkedType(Parameter parameter, boolean isShortType) {
+        final String typeValue;
+        if (isShortType) {
+            typeValue = getShortType(parameter.getType());
+        } else {
+            typeValue = parameter.getType();
+        }
+        URI uri = null;
+        if (parameter instanceof EnhancedParameterWrapper) {
+            EnhancedParameterWrapper enhancedParameter = (EnhancedParameterWrapper) parameter;
+            if (enhancedParameter.getTypeJavadocUrl() != null) {
+                URI javadocUrl = enhancedParameter.getTypeJavadocUrl();
+                // optionally check if link is valid
+                if (javadocUrl.isAbsolute()
+                        || disableInternalJavadocLinkValidation
+                        || JavadocLinkGenerator.isLinkValid(javadocUrl, reportOutputDirectory.toPath())) {
+                    uri = enhancedParameter.getTypeJavadocUrl();
+                }
+            }
+        }
+        // rely on the decoded link (URL encoding is done by the Sink implementation if necessary)
+        Optional<String> link = Optional.ofNullable(uri).map(u -> {
+            StringBuilder decodedLink = new StringBuilder();
+            if (u.getScheme() != null) {
+                decodedLink.append(u.getScheme()).append(":");
+            }
+            decodedLink.append(u.getSchemeSpecificPart());
+            return decodedLink.toString();
```

Comment:
Can you explain why this has to be so complex? Why not convert to string if the URI isn't null?

Reply:
`URI.toString()` returns the URL-encoded String which is 
a) not necessary as the URL is encoded by [the ](https://github.com/apache/maven-doxia/blob/a1e97f9d47b20470c4a5b536406d8c5cb08277bc/doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/Xhtml5BaseSink.java#L1595)
b) double encoding potentially leads to invalid links

Therefore it is crucial that the decoded URI (i.e. unencoded link) is passed to the Sink API!

Reply:
> `URI.toString()` returns the URL-encoded String which is a) not necessary as the URL is encoded by https://github.com/apache/maven-doxia/blob/a1e97f9d47b20470c4a5b536406d8c5cb08277bc/doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/Xhtml5BaseSink.java#L1595 b) double encoding potentially leads to invalid links
> 
> Therefore it is crucial that the decoded URI (i.e. unencoded link) is passed to the Sink API!

I consider both statement as wrong. `URI#toString()` does not encode multibyte chars and there referenced statement just makes sure that the attribute value is valid for XML, the actual content is opaque and may be anything, even `javascript:void()`.

Reply:
URI encodes parameters passed to its constructor for all cases except for the single string argument. Look at the impl. Also look at the description of the getters raw vs non-raw. In general URI internally always stores RFC compliant parts (i.e. properly encoded ones). When using toString() you would URI encode twice! Once via URI itself and the second time via the HTMLSink.

Reply:
> URI encodes parameters passed to its constructor for all cases except for the single string argument. Look at the impl. Also look at the description of the getters raw vs non-raw. In general URI internally always stores RFC compliant parts (i.e. properly encoded ones). When using toString() you would URI encode twice! Once via URI itself and the second time via the HTMLSink.

Where does the sink properly URI-encodes the input? I just don't see it.

Reply:
If the link is handled opaque then it always have to be RFC 2396 in order to use it with the HTML Sink, while other sinks might not need that. That would be a serious flaw in the abstraction IMHO.

Reply:
> If the link is handled opaque then it always have to be RFC 2396 in order to use it with the HTML Sink, while other sinks might not need that. That would be a serious flaw in the abstraction IMHO.

RFC 2396 is dead for a link time so is `URI` broken. The Sink API is backend-agnostic. It cannot know what you are passing. It might be a flaw in the design, but that's the way it is. *You* have to provide valid input.

Reply:
Thanks for the pointer, I hopefully clarified this a bit with https://github.com/apache/maven-doxia/pull/176/files. I am gonna rework it now to use `URI.toASCIIString` to return the URI with proper encoding.

Reply:
Done in https://github.com/apache/maven-plugin-tools/pull/225/commits/1696e357ee40011660c39bfff80d58fa8dfc4ec8.